### ISSUE TITLE
Multi-hop graph relationships

### ISSUE THREAD

**Issue by davidmezzetti** [aplbrain/grand-cypher#35]
In building multi-hop queries, is there currently a method to retrieve the node ids along with the attributes?

i.e for the query

```cypher
MATCH (A{id: "Vikings"})-[R*0..3]->(B{id: "England"})
RETURN A, R, B
LIMIT 1
```

The source and target nodes could be added as attributes. In looking at the code, it's also straightforward to add those in all cases or with multi-hop relationships. Lastly, the openCypher standard has `startNode` and `endNode` functions. 

Curious if there are any thoughts on this use case.

**Comment by j6k4m8** [comment]
I think we have a lot of the heavy lifting already set up for this (#24), but we don't currently implement it!

Linking #23 and #26 for cross-reference!

**Comment by davidmezzetti** [comment]
Interesting, I'll take a look. I'm happy to submit a PR for this feature. It sounds like you'd like it to behave like what was proposed in #26?

**Comment by j6k4m8** [comment]
That would be so awesome!! I'd need to take a look at the openCypher spec again; the ideal functionality is 1-to-1 feature parity with how this is handled in existing implementations. Which I _think_ is what we mentioned in #26...

**Comment by davidmezzetti** [comment]
Sounds great.

In reading the openCypher spec here - https://s3.amazonaws.com/artifacts.opencypher.org/openCypher9.pdf

And it looks like for a query like this (page 65).

```cypher
MATCH p = (charlie:Person)-[* {blocked:false}]-(martin:Person)
WHERE charlie.name = 'Charlie Sheen' AND martin.name = 'Martin Sheen'
RETURN p
```

It returns a list like this for each row.

```python
[Node[0]{name:"Charlie
Sheen"},:X[20]{blocked:false},Node[20]{},:X[21]{blocked:false},Node[1]{name:"Martin Sheen"}]
```

I think most of the work is already done for this. It looks to just be a matter of changing the parser spec to handle the path definition `<var> = ` and returning the path. In looking at the example though, it's hard to see how that maps to a Python structure. Is it a list of single element dictionaries?

**Comment by j6k4m8** [comment]
Hm weird — yeah, I'd interpret that as an iterable of Vertex elements, like

```python
[
# Node
{id: 0, name:"Charlie Sheen"},
# Edge:X
{id: 20, blocked:false},
# Node
{id: 20},
# Edge:X
{id: 21, blocked:false},
# Node
{id: 1, name:"Martin Sheen"}
]
```

There is the issue that we don't really have a good way of distinguishing vertex and edge entities as dictionaries... In our graph repr, we have the `__labels__` attr that we're more or less standardizing kinda just by virtue of using it in a lot of places now (https://github.com/aplbrain/grand-cypher/pull/25) ([gc-io](https://github.com/aplbrain/grand-cypher-io?tab=readme-ov-file#the-__labels__-magic-attribute)), which is maybe a way to do this...

```python
{id: 1, name:"Martin Sheen", __labels__=["Vertex"]}
```
though I _really_ don't like the idea of a vertex having a different number of attributes than it has in the db...

Curious what you think @davidmezzetti! Also tagging @khoale88 who has had a lot of good insight on this in the past :)

**Comment by davidmezzetti** [comment]
Adding an attribute that notes the type seems to make sense to me.

**Comment by khoale88** [comment]
I haven't touched this for quiet sometime, but luckily I'm still be able to recall something. Here are my 2 cents. The decision can depend on how the information is used. 
- If we don't want to touch the core code for now and accessing path entity type is not mandatory, a custom structure can work as the 1st iteration. Something like this would be enough for the human readability.
```python
class NodePathView(dict):
  def __repr__(self):
    return f"Node{super().__repr__()}"
```
- To go further, I think that putting additional attribute in the labels might not be as friendly as having a separated attributes, both in code clarity and performance wise. I perfer to have a dedicated attribute for this, since this attribute is static. Though, I rekon that reusing labels can save us some development effort. Using labels is definitely a good beginning. One note is that try to make it the same for both vertex and edge so that we don't have to remember 2 different things.

**Comment by davidmezzetti** [comment]
Fixed with #37
